rpc: Enable wallet import on pruned nodes and add test

aureleoules opened the pull request:
Reopens #16037 

I have rebased the PR, addressed the comments of the original PR and added a functional test.

> Before this change importwallet fails if any block is pruned. This PR makes it possible to importwallet if all required blocks aren't pruned. This is possible because the dump format includes key timestamps.

For reviewers: 
`python test/functional/wallet_pruning.py --nocleanup` will generate a large blockchain (~700MB) that can be used to manually test wallet imports on a pruned node. Node0 is not pruned, while node1 is.

vincenzopalazzo commented on test/functional/wallet_pruning.py:
```diff
@@ -0,0 +1,161 @@
+    assert_raises_rpc_error,
+)
+
+from test_framework.blocktools import (create_coinbase)
```
```suggestion
from test_framework.blocktools import create_coinbase
```

vincenzopalazzo commented on test/functional/wallet_pruning.py:
```diff
@@ -0,0 +1,161 @@
+    get_key,
+)
+from test_framework.test_framework import BitcoinTestFramework
+
+class WalletPruningTest(BitcoinTestFramework):
```
```suggestion


class WalletPruningTest(BitcoinTestFramework):
```

Usualy in python there are two spaces between class and the last import declaration

Xekyo commented on test/functional/wallet_pruning.py:
```diff
@@ -0,0 +1,157 @@
+class WalletPruningTest(BitcoinTestFramework):
+    nTime = 0
+    def mine_large_blocks(self, node, n):
+        # Taken from test/functional/feature_pruning.py
```
If we're reusing helper functions, perhaps we should extract them to a single place in a follow-up PR.

murchandamus reviewed (commented):
Did a light code-review, concept ACK

achow101 commented on src/wallet/rpc/backup.cpp:
```diff
@@ -1383,6 +1388,8 @@ RPCHelpMan importmulti()
             }
         }
+
+        EnsureBlockDataFromTime(pwallet->chain(), nLowestTimestamp);
```
In 46b7f731b1e98b0836c5b97cd99c0887d6173a17 "rpc: Enable wallet import on pruned nodes and add test"

We don't want to do this for `importmulti`. `importmulti` can have some imports succeed, and others fail. Just because one import has an out of range timestamp does not mean the entire import should fail the RPC. This check would result in a JSONRPCError which is not what we expect from `importmulti` - rather there should be an error per import that has a too old timestamp.

Furthermore, we already handle the pruned node case. If the rescan reports a time that is more recent than the timestamp for a given import, we will give an error for that import with more details about what happened, maybe why it happened, and how the user can resolve it.

kouloumos reviewed (commented):
I think that `wallet_pruning.py` is not testing properly this added feature. This is due to the testing logic but mainly because of the default setup of the testing framework that is being used in this test. 

> This is possible because the dump format includes key timestamps.

The `importwallet` rpc calculates the min `nTimeBegin` timestamp for which `findFirstBlockWithTimeAndHeight` in the newly created function `EnsureBlockDataFromTime` will find the earliest block with timestamp equal or greater to.

As you can verify from the `wallet_*.dat` dumps of this test
```
# node1 wallet dump
...
cUxsWyKyZ9MAQTaAhUQWJmBbSvHMwSmuv59KgxQV7oZQU3PXN3KE 1970-01-01T00:00:01Z label=coinbase # addr=msX6jQXvxiNhx3Q62PKeLPrhrqZQdSimTg,2MtBq31Chq2pntJ81PQLR4q4sMEG7K46wsQ,bcrt1qsw5g6ehh439vurfyhdh93d66hw0kf9085wy7ma
...
```
the deterministic private keys that are imported to each test node as part of the testing framework setup 
https://github.com/bitcoin/bitcoin/blob/1d89fc695a3aeb3e3dcadf371b7667572b38c836/test/functional/test_framework/test_framework.py#L441

have a `1970-01-01T00:00:01Z` timestamp which means that `EnsureBlockDataFromTime` will always check for the existence of all blocks between genesis and the current height. 

Even if that was not the case, because the nodes' wallets are initialized before we start generating blocks, the timestamp in the wallet dump is still before the first created block thus `EnsureBlockDataFromTime` checks for the existence of all blocks between the first block and the current height. 

This behaviour will always result to missing blocks as soon as we start pruning. The reason that the first part of the test (`wallet_import_pruned_test`) is not failing is because it never reaches the prune state (height < PruntAfterHeight).

I believe that you can better test for the new feature by overriding `setup_nodes` and initializing node1 wallet mid-test

aureleoules commented:
Thank you for your thorough review @kouloumos !
I believe I have addressed all of your comments.

kouloumos commented on src/wallet/rpc/backup.cpp:
```diff
@@ -93,6 +93,20 @@ static void RescanWallet(CWallet& wallet, const WalletRescanReserver& reserver,
+    const std::optional<int> chainHeight{chain.getHeight()};
+
+    if (found && chainHeight && !chain.hasBlocks(chain.getBlockHash(chainHeight.value()), height)) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "Pruned blocks required to import keys");
```
I think that at this point we have the required `height` for which we need blocks. Could this change to a more descriptive error message stating that required weight? And maybe "[...]. Use RPC call getblockchaininfo to determine your pruned height" as [here](https://github.com/bitcoin/bitcoin/blob/15692e2641592394bdd4da0a7c2d371de8e576dd/src/wallet/rpc/transactions.cpp#L898)?

aureleoules replied:
thanks, added it

ryanofsky reviewed (approved):
Code review ACK 0aedf6f3cb5a513f9e42458362261d4afa2ea17f. Nice change and tests!

aureleoules commented on test/functional/test_framework/blocktools.py:
```diff
@@ -87,6 +87,13 @@ def create_block(hashprev=None, coinbase=None, ntime=None, *, version=None, tmpl
     block.calc_sha256()
     return block
 
+def create_large_block(hashprev=None, ntime=None, height=None, scriptPubKey=CScript([OP_RETURN] + [OP_TRUE] * 950000)):
```
This is a large object that takes time to construct so I thought making it static would avoid constructing it multiple times, but I had not thought of the fact that this lives in memory even if the function is unused.
Will fix though, thanks.

aureleoules commented on test/functional/feature_index_prune.py:
```diff
@@ -30,10 +30,10 @@ def sync_index(self, height):
             'coinstatsindex': {'synced': True, 'best_block_height': height}
         }
-        self.wait_until(lambda: self.nodes[1].getindexinfo() == expected_stats)
+        self.wait_until(lambda: self.nodes[1].getindexinfo() == expected_stats, timeout=120)
```
oops, fixed

aureleoules commented on src/wallet/rpc/backup.cpp:
```diff
@@ -93,6 +93,20 @@ static void RescanWallet(CWallet& wallet, const WalletRescanReserver& reserver,
+    const bool found{chain.findFirstBlockWithTimeAndHeight(timestamp - TIMESTAMP_WINDOW, 0, FoundBlock().height(height))};
+    const std::optional<int> chainHeight{chain.getHeight()};
+
+    if (found && chainHeight && !chain.hasBlocks(chain.getBlockHash(chainHeight.value()), height)) {
```
Thanks for the review, fixed it!

ryanofsky commented on src/wallet/rpc/backup.cpp:
```diff
@@ -93,6 +93,21 @@ static void RescanWallet(CWallet& wallet, const WalletRescanReserver& reserver,
+
+    int height{0};
+    const bool found{chain.findFirstBlockWithTimeAndHeight(timestamp - TIMESTAMP_WINDOW, 0, FoundBlock().height(height))};
+    const std::optional<int> chainHeight{chain.getHeight()};
```
In commit "rpc: Enable wallet import on pruned nodes and add test" (f42c077a8f948f5c93b68eb1b7fa246b4f6475ef)

I think there's no need for the chainHeight variable anymore and it would be better to drop it.

ryanofsky commented on test/functional/feature_pruning.py:
```diff
@@ -10,25 +10,25 @@
 # compatible with pruning based on key creation time.
 TIMESTAMP_WINDOW = 2 * 60 * 60
 
+scriptPubKey = CScript([OP_RETURN] + [OP_TRUE] * 950000)
```
In commit "rpc: Enable wallet import on pruned nodes and add test" (f42c077a8f948f5c93b68eb1b7fa246b4f6475ef)

Just IMO, but I think it would be better to keep the existing `big_script` variable below and delete this new variable (smaller diff, more descriptive variable name, more straightforward code avoiding global variables)

ryanofsky commented on test/functional/test_framework/blocktools.py:
```diff
@@ -87,6 +87,13 @@ def create_block(hashprev=None, coinbase=None, ntime=None, *, version=None, tmpl
     block.calc_sha256()
     return block
 
+def create_large_block(hashprev, ntime, height, scriptPubKey):
```
In commit "rpc: Enable wallet import on pruned nodes and add test" (f42c077a8f948f5c93b68eb1b7fa246b4f6475ef)

This new `create_large_block` seems very inflexible and also has a misleading name. Would drop it and instead just add a new option to `create_coinbase` like:

```diff
diff --git a/test/functional/feature_pruning.py b/test/functional/feature_pruning.py
index 8d72d11baf5..d14970c9463 100755
--- a/test/functional/feature_pruning.py
+++ b/test/functional/feature_pruning.py
@@ -10,7 +10,8 @@ This test takes 30 mins or more (up to 2 hours)
 """
 import os
 
-from test_framework.blocktools import create_large_block
+from test_framework.blocktools import create_block
+from test_framework.blocktools import create_coinbase
 from test_framework.test_framework import BitcoinTestFramework
 from test_framework.util import (
     assert_equal,
@@ -47,7 +48,7 @@ def mine_large_blocks(node, n):
     previousblockhash = int(best_block["hash"], 16)
 
     for _ in range(n):
-        block = create_large_block(hashprev=previousblockhash, height=height, ntime=mine_large_blocks.nTime, scriptPubKey=scriptPubKey)
+        block = create_block(hashprev=previousblockhash, ntime=mine_large_blocks.nTime, coinbase=create_coinbase(height, script_pubkey=scriptPubKey))
         block.solve()
 
         # Submit to the node
diff --git a/test/functional/test_framework/blocktools.py b/test/functional/test_framework/blocktools.py
index 5916ffc2b17..77653df96e9 100644
--- a/test/functional/test_framework/blocktools.py
+++ b/test/functional/test_framework/blocktools.py
@@ -87,13 +87,6 @@ def create_block(hashprev=None, coinbase=None, ntime=None, *, version=None, tmpl
     block.calc_sha256()
     return block
 
-def create_large_block(hashprev, ntime, height, scriptPubKey):
-    coinbase_tx = create_coinbase(height)
-    coinbase_tx.vout[0].scriptPubKey = scriptPubKey
-    coinbase_tx.rehash()
-
-    return create_block(coinbase=coinbase_tx, ntime=ntime, hashprev=hashprev)
-
 def get_witness_script(witness_root, witness_nonce):
     witness_commitment = uint256_from_str(hash256(ser_uint256(witness_root) + ser_uint256(witness_nonce)))
     output_data = WITNESS_COMMITMENT_HEADER + ser_uint256(witness_commitment)
@@ -127,7 +120,7 @@ def script_BIP34_coinbase_height(height):
     return CScript([CScriptNum(height)])
 
 
-def create_coinbase(height, pubkey=None, extra_output_script=None, fees=0, nValue=50):
+def create_coinbase(height, pubkey=None, *, script_pubkey=None, extra_output_script=None, fees=0, nValue=50):
     """Create a coinbase transaction.
 
     If pubkey is passed in, the coinbase output will be a P2PK output;
@@ -145,6 +138,8 @@ def create_coinbase(height, pubkey=None, extra_output_script=None, fees=0, nValu
         coinbaseoutput.nValue += fees
     if pubkey is not None:
         coinbaseoutput.scriptPubKey = key_to_p2pk_script(pubkey)
+    elif script_pubkey is not None:
+        coinbaseoutput.scriptPubKey = script_pubkey
     else:
         coinbaseoutput.scriptPubKey = CScript([OP_TRUE])
     coinbase.vout = [coinbaseoutput]
diff --git a/test/functional/wallet_pruning.py b/test/functional/wallet_pruning.py
index 9b8160a4490..e42994422e0 100755
--- a/test/functional/wallet_pruning.py
+++ b/test/functional/wallet_pruning.py
@@ -7,7 +7,8 @@
 import os
 
 from test_framework.util import assert_raises_rpc_error
-from test_framework.blocktools import create_large_block
+from test_framework.blocktools import create_block
+from test_framework.blocktools import create_coinbase
 from test_framework.test_framework import BitcoinTestFramework
 
 from test_framework.script import (
@@ -26,7 +27,7 @@ class WalletPruningTest(BitcoinTestFramework):
         self.nTime = max(self.nTime, int(best_block["time"])) + 1
         previousblockhash = int(best_block["hash"], 16)
         for _ in range(n):
-            block = create_large_block(hashprev=previousblockhash, height=height, ntime=self.nTime, scriptPubKey=self.scriptPubKey)
+            block = create_block(hashprev=previousblockhash, ntime=self.nTime, coinbase=create_coinbase(height, script_pubkey=self.scriptPubKey))
             block.solve()
 
             # Submit to the node

```

ryanofsky commented on src/wallet/rpc/backup.cpp:
```diff
@@ -93,6 +93,21 @@ static void RescanWallet(CWallet& wallet, const WalletRescanReserver& reserver,
+    const bool found{chain.findFirstBlockWithTimeAndHeight(timestamp - TIMESTAMP_WINDOW, 0, FoundBlock().height(height))};
+    const std::optional<int> chainHeight{chain.getHeight()};
+
+    LOCK(wallet.cs_wallet);
```
In commit "rpc: Enable wallet import on pruned nodes and add test" (f42c077a8f948f5c93b68eb1b7fa246b4f6475ef)

I don't think we need to keep the wallet locked while running all the code below. Would suggest using a shorter lock:

```c++
uint256 tip_hash = WITH_LOCK(wallet.cs_wallet, return wallet.GetLastBlockHash());
```

ryanofsky reviewed (approved):
Code review ACK f42c077a8f948f5c93b68eb1b7fa246b4f6475ef. Just suggested changes since last review, and increasing sync_blocks timeout in the new test. Thanks for the updates!

aureleoules commented:
Thanks for the review again @ryanofsky, I think I've addressed all your comments.

ryanofsky commented on test/functional/feature_pruning.py:
```diff
@@ -41,28 +41,14 @@ def mine_large_blocks(node, n):
     height = int(best_block["height"]) + 1
     mine_large_blocks.nTime = max(mine_large_blocks.nTime, int(best_block["time"])) + 1
     previousblockhash = int(best_block["hash"], 16)
+    bigScript = CScript([OP_RETURN] + [OP_TRUE] * 950000)
```
In commit "rpc: Enable wallet import on pruned nodes and add test" (05463d4001001213c17672c1607a3a493927ddd7)

Can just keep existing `big_script` variable, no need to replace it with new `bigScript` variable. This would make the diff small and make the naming convention more convenient

aureleoules replied:
my bad, fixed

ryanofsky reviewed (approved):
Code review ACK 05463d4001001213c17672c1607a3a493927ddd7. Only changes since last review were some suggested cleanups & simplifications

ryanofsky reviewed (approved):
Code review ACK 6dc2e7141de83022f0ca0c48afd5ea7e5c4684d0. Only change since last review is changing bigScript back to big_script

w0xlt reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/24865/commits/6dc2e7141de83022f0ca0c48afd5ea7e5c4684d0

aureleoules commented:
Fixes #9409.

kouloumos commented on test/functional/wallet_pruning.py:
```diff
@@ -0,0 +1,132 @@
+
+        self.create_big_chain()
+
+        # connect mid-test node1
```
nit: maybe add some comments here to explain the rationale behind this mid-test connection? I think it will not be very clear for someone reading this test in the future.

kouloumos commented on test/functional/wallet_pruning.py:
```diff
@@ -0,0 +1,132 @@
+
+        wname = "wallet_pruned.dat"
+
+        # export wallet
+        self.nodes[1].dumpwallet(os.path.join(self.nodes[1].datadir, wname))
+        # import wallet
+        self.nodes[1].importwallet(os.path.join(self.nodes[1].datadir, wname))
```
This dump&import pattern in the same wallet might be ok for checking against the RPC assertion of the newly introduced logic, but in reality the keys and scripts in the dump file are not imported in the wallet because they already exist. 

I don't consider this as an issue because that behavior is also tested elsewhere(`wallet_dump.py`) but I think changing this part makes for a better usage scenario.

kouloumos reviewed (commented):
reACK 6dc2e7141de83022f0ca0c48afd5ea7e5c4684d0

I've also verified the expected behavior on a pruned signet node by dumping the wallet and then manually changing the timestamps in the file to simulate older keys for which the block data do not exist (pruned).

Does it make sense to also add a check (in the functional tests), that after restoring, the wallet correctly accounts for balances and maybe addresses? Or that's redundant because a similar check exists at `wallet_dump.py`?
https://github.com/bitcoin/bitcoin/blob/33eef562a321ce772a9a88073a78a85894cb3fe8/test/functional/wallet_dump.py#L198-L206

aureleoules committed:
rpc: Enable wallet import on pruned nodes

Co-authored-by: João Barbosa <joao.paulo.barbosa@gmail.com>
Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

aureleoules committed:
test: Wallet imports on pruned nodes

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>
Co-authored-by: Andreas Kouloumos <kouloumosa@gmail.com>

aureleoules commented:
Thank you @kouloumos for your review.
I have reworked the test to take your comments into account and also reduce the time it takes to run.
It took 1min24 to run and reduced it to 54 seconds (on my machine).
Also added more comments to explain each step of the test and split the PR into two commits.

aureleoules commented:
Added a test case to check that the wallet balance is correct.

andrewtoth commented:
Sorry, I didn't check that https://github.com/bitcoin/bitcoin/pull/26206 will actually conflict with this PR. This needs a rebase on top of it and then revert that commit in this PR.

aureleoules commented:
Rebased, reverted 4aff7a48a4e0f1075306f181a276b8a74c857022 and added @ryanofsky and @kouloumos as co-authors.

andrewtoth commented on src/wallet/rpc/backup.cpp:
```diff
@@ -93,6 +93,20 @@ static void RescanWallet(CWallet& wallet, const WalletRescanReserver& reserver,
     }
 }
 
+static void EnsureBlockDataFromTime(interfaces::Chain& chain, const CWallet& wallet, int64_t timestamp)
```
Why make a new function instead of doing this inline?

andrewtoth commented on src/wallet/rpc/backup.cpp:
```diff
@@ -93,6 +93,20 @@ static void RescanWallet(CWallet& wallet, const WalletRescanReserver& reserver,
     }
 }
 
+static void EnsureBlockDataFromTime(interfaces::Chain& chain, const CWallet& wallet, int64_t timestamp)
```
Why create a function for this? Would this be easier to reason about if it was inline?

aureleoules replied:
I've seen many instances of `Ensure(.*)` functions and it also helps to reduce the length of the parent function I suppose.

kouloumos commented:
Does this need a rebase after https://github.com/bitcoin/bitcoin/pull/26380?

Smithbull001 reviewed (commented):
Enable wallet connect on pruned network

MarcoFalke commented on src/wallet/rpc/backup.cpp:
```diff
@@ -93,6 +93,20 @@ static void RescanWallet(CWallet& wallet, const WalletRescanReserver& reserver,
 
+static void EnsureBlockDataFromTime(interfaces::Chain& chain, const CWallet& wallet, int64_t timestamp)
+{
+    if (!chain.havePruned())
```
```suggestion
    if (!chain.havePruned()) {
```

nit (for new code)

aureleoules commented:
Dropped e1eadaa72d6831d1d0a53ba97c215dc4cdb64436, rebased for #26380 and addressed MarcoFalke's nit.

achow101 commented on test/functional/wallet_pruning.py:
```diff
@@ -0,0 +1,148 @@
+        # Generate 50MB worth of large blocks in the second blk*.dat file
+        self.mine_large_blocks(self.nodes[0], 50)
+        # Create a wallet and a backup of it which birth's block is in the second blk*.dat file
+        self.nodes[0].createwallet(wallet_name="wallet_init", descriptors=self.options.descriptors, load_on_startup=True)
```
In 6abd81b11dd1a45bdffcc85ba90591474a3f48b0 "test: Wallet imports on pruned nodes"

Since this is testing legacy wallet only RPCs, we should not be looking at the `--legacy-wallet`/`--descriptors` options to determine the wallet type to create, for all wallets created in this test.

```suggestion
        self.nodes[0].createwallet(wallet_name="wallet_init", descriptors=False, load_on_startup=True)
```

achow101 commented on test/functional/test_runner.py:
```diff
@@ -83,6 +83,7 @@
     'feature_pruning.py',
     'feature_dbcrash.py',
     'feature_index_prune.py',
+    'wallet_pruning.py',
```
In 6abd81b11dd1a45bdffcc85ba90591474a3f48b0 "test: Wallet imports on pruned nodes"

Should have `--legacy-wallet` to indicate that this test is a legacy wallet only test. Also, the default is `--descriptors`.

```suggestion
    'wallet_pruning.py --legacy-wallet',
```

w0xlt commented on test/functional/feature_pruning.py:
```diff
@@ -10,8 +10,8 @@
 """
 import os
 
+from test_framework.blocktools import create_block
```
from test_framework.blocktools import (
    create_block,
    create_coinbase,
)

w0xlt replied:
```suggestion
from test_framework.blocktools import (
    create_block,
    create_coinbase,
)
```

w0xlt commented on test/functional/feature_pruning.py:
```diff
@@ -10,8 +10,8 @@
 """
 import os
 
+from test_framework.blocktools import create_block
 from test_framework.blocktools import create_coinbase
```
```suggestion
from test_framework.blocktools import (
    create_block,
    create_coinbase,
)
```
```

w0xlt reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/24865/commits/6abd81b11dd1a45bdffcc85ba90591474a3f48b0

aureleoules commented:
Addressed the suggestions, thanks @achow101 @w0xlt.

achow101 commented on test/functional/wallet_pruning.py:
```diff
@@ -0,0 +1,148 @@
+        wname = "wallet_pruned.dat"
+
+        # Export wallet from node 0
+        self.nodes[0].createwallet(wallet_name="wallet_pruned", descriptors=self.options.descriptors, load_on_startup=True)
```
In a0762cf6ec0bbc5eb770912a1ab917862f5515ca "test: Wallet imports on pruned nodes"

```suggestion
        self.nodes[0].createwallet(wallet_name="wallet_pruned", descriptors=False, load_on_startup=True)
```

aureleoules replied:
Oops, pushed.

achow101 commented on test/functional/wallet_pruning.py:
```diff
@@ -0,0 +1,148 @@
+        self.generate(self.nodes[0], 100)
+
+        # Import wallet into node 1
+        self.nodes[1].createwallet(wallet_name="wallet_pruned", descriptors=self.options.descriptors, load_on_startup=True)
```
In a0762cf6ec0bbc5eb770912a1ab917862f5515ca "test: Wallet imports on pruned nodes"

```suggestion
        self.nodes[1].createwallet(wallet_name="wallet_pruned", descriptors=False, load_on_startup=True)
```

andrewtoth commented:
You should rebase to fix ci test failure for `p2p_sendtxrcncl.py`.

aureleoules commented:
> You should rebase to fix ci test failure for p2p_sendtxrcncl.

:heavy_check_mark:

furszy commented on test/functional/wallet_pruning.py:
```diff
@@ -0,0 +1,148 @@
+    def test_wallet_import_pruned_with_missing_blocks(self):
+        self.log.info("Make sure we cannot import wallet when pruned and required blocks are not available")
+
+        wname = "wallet_init.dat"
```
In c46f5fe5:

Tiny nit, this could be passed as an argument so we avoid hardcoding file names more than once.

furszy reviewed (approved):
Code ACK c46f5fe5

theStack reviewed (commented):
Concept ACK

aureleoules commented:
Addressed the nits, thanks @furszy @theStack.
https://github.com/bitcoin/bitcoin/compare/c46f5fe58e60ad1ff25978604243a90fce6b634c..eb2e54d00e7c03e0ca5238a2fabe078fca08b1c2

aureleoules commented:
Reduced the number of blocks generated to speed up the test (from 1:40 to 1:24).

kouloumos reviewed (commented):
The functional test got me really confused on what it is testing and how it is testing it. It might be because I am re-reviewing this after 55 days, but I think that some simplification (left comments inline) as well as commenting on the basis of the change and not on what the code does could help with that.

What the test does (afaiu) is simulating 2 different wallets with 2 different birth heights
`wallet_birthheight_1` < `wallet_birthheight_2`
It then prunes below `birthheight_2` to:

- check with `test_wallet_import_pruned_test` that `wallet_birthheight_2` is imported correctly.
- check with `test_wallet_import_pruned_with_missing_blocks` that `wallet_birthheight_1` cannot imported correctly.

Based on the test overview above and some comments that I left inline, I did some implementation-logic restructuring and added a few comments that I think can help with comprehension. You can find my changes at 69b9b6247305d06168f58e9e156807e09a925bf6 ([diff](https://github.com/bitcoin/bitcoin/compare/c594dbc7560f18a657bd624983ca58732b0f3183...69b9b6247305d06168f58e9e156807e09a925bf6)), feel free to pick whatever you like.

aureleoules committed:
test: Introduce MIN_BLOCKS_TO_KEEP constant

aureleoules commented on src/wallet/rpc/backup.cpp:
```diff
@@ -565,15 +573,18 @@ RPCHelpMan importwallet()
             }
         }
         file.close();
+        EnsureBlockDataFromTime(pwallet->chain(), *pwallet, nTimeBegin);
```
Removed chain argument, thanks.

aureleoules commented on test/functional/wallet_pruning.py:
```diff
@@ -0,0 +1,150 @@
+
+        # A blk*.dat file is 128MB
+        # Generate 288 light blocks, the minimum required to stay on disk with prune enabled
+        self.generate(self.nodes[0], MIN_BLOCKS_TO_KEEP, sync_fun=lambda: None)
```
Updated thanks.

aureleoules commented on test/functional/wallet_pruning.py:
```diff
@@ -0,0 +1,150 @@
+        self.generate(self.nodes[0], MIN_BLOCKS_TO_KEEP, sync_fun=lambda: None)
+
+        # Generate 50MB worth of large blocks in the second blk*.dat file
+        self.mine_large_blocks(self.nodes[0], 50)
```
Yes the comment was wrong.

aureleoules commented on test/functional/wallet_pruning.py:
```diff
@@ -0,0 +1,150 @@
+
+        # A blk*.dat file is 128MB
+        # Generate 288 light blocks, the minimum required to stay on disk with prune enabled
+        self.generate(self.nodes[0], MIN_BLOCKS_TO_KEEP, sync_fun=lambda: None)
```
I think this was based on a different version of this test but you are correct 288 is not needed. I changed it to 250.

aureleoules replied:
I think this was based on a different version of this test but you are correct 288 is not needed. I changed it to 250.

aureleoules commented on test/functional/wallet_pruning.py:
```diff
@@ -0,0 +1,150 @@
+
+class WalletPruningTest(BitcoinTestFramework):
+    nTime = 0
+    def mine_large_blocks(self, node, n):
```
Thanks, I moved it.

aureleoules commented on test/functional/wallet_pruning.py:
```diff
@@ -0,0 +1,150 @@
+)
+
+class WalletPruningTest(BitcoinTestFramework):
+    nTime = 0
```
Thanks I moved it.

aureleoules commented:
Thanks for your detailed review @kouloumos. I cherry-picked your commit and squashed it. I also addressed some of your additional suggestions.

kouloumos commented:
ACK 8e87433a4460088d15275623f0b6f1893af28d82
nit: The `MIN_BLOCKS_TO_KEEP` constant is not used anymore in the test, which makes 8e87433a4460088d15275623f0b6f1893af28d82 out of scope.

MarcoFalke commented:
This is adding a new test `wallet_pruning.py`, but I can't find it in the CI output at all. Does anyone see what is going on here?

fanquake commented:
Is it becuase it's in EXTENDED_SCRIPTS?

kouloumos commented:
It also runs as part of [Win64 native [vs2022]](https://github.com/bitcoin/bitcoin/pull/24865/checks?check_run_id=9728080955). I can see that the extended tests run there and `wallet_pruning.py` runs successfully.

MarcoFalke commented:
Ah thx. Maybe there should be a `skip_if_not_running_extended` :sweat_smile: ?


Also, I think this needs a rebase, see upcoming CI failure in https://cirrus-ci.com/task/6513158118965248

aureleoules commented:
Rebased to address silent conflicts. The current CI error is unrelated.

achow101 commented:
ACK 63493d778f443554059d77bd8c9b1a926685a892

kouloumos commented on test/functional/wallet_pruning.py:
```diff
@@ -0,0 +1,158 @@
+
+class WalletPruningTest(BitcoinTestFramework):
+    def add_options(self, parser):
+        self.add_wallet_options(parser)
```
If this is a legacy wallet only test, this should be
```suggestion
        self.add_wallet_options(parser, descriptors=False)
```

kouloumos commented on test/functional/wallet_pruning.py:
```diff
@@ -0,0 +1,158 @@
+
+    def run_test(self):
+        self.nTime = 0
+        self.log.info("Warning! This test requires ~1.1GB of disk space")
```
nit I haven't calculate it properly, but with some quick monitoring during the test run, it seems that this might be closer to 1.3GB

kouloumos commented:
ACK 564b580bf07742483a140c7c095b896a6d5d6cad

furszy commented on test/functional/test_framework/blocktools.py:
```diff
@@ -138,6 +138,8 @@ def create_coinbase(height, pubkey=None, extra_output_script=None, fees=0, nValu
         coinbaseoutput.nValue += fees
     if pubkey is not None:
         coinbaseoutput.scriptPubKey = key_to_p2pk_script(pubkey)
+    elif script_pubkey is not None:
+        coinbaseoutput.scriptPubKey = script_pubkey
```
In 71d9a7c0:

atm is just a non-blocking nit:

it's weird to add this new ´script_pubkey´ arg that is only set if the ´pubkey´ arg is not provided and not stated anywhere.

Ideally, should unify both into a single arg. So there is no confusion over which of them take precedence over the other.

furszy commented on test/functional/wallet_pruning.py:
```diff
@@ -0,0 +1,158 @@
+            # Important for matching a timestamp with a block +- some window
+            self.nTime += 600
+            for n in self.nodes:
+                if n.running:
```
In https://github.com/bitcoin/bitcoin/commit/71d9a7c03b44236c2fea2b74f92a69234d29f717:
why the node would not be running?

aureleoules replied:
Right, in a previous version of the test it could crash without this check.

furszy reviewed (approved):
ACK 564b580
left few non-blocking nits.

w0xlt reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/24865/commits/564b580bf07742483a140c7c095b896a6d5d6cad

achow101 commented:
reACK 564b580bf07742483a140c7c095b896a6d5d6cad

MarcoFalke commented on test/functional/wallet_pruning.py:
```diff
@@ -0,0 +1,158 @@
+        previousblockhash = int(best_block["hash"], 16)
+        big_script = CScript([OP_RETURN] + [OP_TRUE] * 950000)
+        for _ in range(n):
+            block = create_block(hashprev=previousblockhash, ntime=self.nTime, coinbase=create_coinbase(height, script_pubkey=big_script))
```
nit: It is possible to do this without blocktools. See

```diff
diff --git a/test/functional/wallet_pruning.py b/test/functional/wallet_pruning.py
index 6d8475ce8d..d768130776 100755
--- a/test/functional/wallet_pruning.py
+++ b/test/functional/wallet_pruning.py
@@ -40,20 +40,10 @@ class WalletPruningTest(BitcoinTestFramework):
     def mine_large_blocks(self, node, n):
         # Get the block parameters for the first block
         best_block = node.getblock(node.getbestblockhash())
-        height = int(best_block["height"]) + 1
         self.nTime = max(self.nTime, int(best_block["time"])) + 1
-        previousblockhash = int(best_block["hash"], 16)
-        big_script = CScript([OP_RETURN] + [OP_TRUE] * 950000)
+        big_script = "raw({})".format(bytes([OP_RETURN] + [OP_TRUE] * 950000).hex())
         for _ in range(n):
-            block = create_block(hashprev=previousblockhash, ntime=self.nTime, coinbase=create_coinbase(height, script_pubkey=big_script))
-            block.solve()
-
-            # Submit to the node
-            node.submitblock(block.serialize().hex())
-
-            previousblockhash = block.sha256
-            height += 1
-
+            self.generateblock(node, big_script, [])
             # Simulate 10 minutes of work time per block
             # Important for matching a timestamp with a block +- some window
             self.nTime += 600
```

Also, might be good to double check the runtime in light of https://github.com/bitcoin/bitcoin/commit/6c872d5e656a7117bbdf19a0220572b93de16f31
